### DISCUSSION TITLE
Ising gauge theory with NetKet

### DISCUSSION THREAD

**Discussion by abukva** [netket/netket#1329] [Q&A] [answered]
I wanted to implement a simple 2D Ising gauge theory, which is just a bunch of Ising degrees of freedom living on a links of 2D lattice:
$H_{Z_2}=-\sum_i \sigma_i^x - \lambda \sum_{\square} \sigma^z \sigma^z \sigma^z \sigma^z$ where $\square$ is sum over all single square plaquettes. This theory is dual to a standard 2D quantum Ising $-\lambda \sum_i \sigma_i^x - \sum_{\langle ij \rangle} \sigma_i^z \sigma_j^z$, so my logic is to test gauge theory results by comparing it to an Ising which I know how to implement. In a conversation with @PhilipVinc he suggested creating Hilber space with number of degrees of freedom same as the number of edges on a lattice and then manually connecting desired links.

```
N = 3
lam = 2.
g = nk.graph.Square(N, pbc=True)

hi = nk.hilbert.Spin(0.5)**(g.n_edges)
ha = nk.operator.LocalOperator(hi)

for i in range(g.n_edges):
    ha -= nk.operator.spin.sigmax(hi,i)

for i in range(N):
    i_next = (i+1) % N
    for j in range(N):
        j_next = (j+1) % N
        ha = ha - lam * nk.operator.spin.sigmaz(hi,edgeIndex(linearIndex(i,j),0))* \
        nk.operator.spin.sigmaz(hi,edgeIndex(linearIndex(i_next,j),1))* \
        nk.operator.spin.sigmaz(hi,edgeIndex(linearIndex(i,j_next),0))* \
        nk.operator.spin.sigmaz(hi,edgeIndex(linearIndex(i,j),1))

ma = nk.models.RBM(alpha=1)
sa = nk.sampler.MetropolisLocal(hilbert=hi, n_chains_per_rank=1)
op = nk.optimizer.Sgd(learning_rate=0.05)
sr = nk.optimizer.SR(diag_shift=0.1)
vs = nk.vqs.MCState(sa, ma, n_samples_per_rank=4000, n_discard_per_chain=100)

gs = nk.VMC(hamiltonian=ha,
            optimizer=op,
            preconditioner=sr,
            variational_state=vs)
gs.run(out=None, n_iter=100)
```
`edgeIndex` and `linearIndex` are just two function that will give me the correct index for a desired link. Once I run this code the energy of a ground state doesn't seem to match with the 2D Ising and also $\sigma^2$ of that estimation is huge, so clearly something is wrong with the convergence. One thing that comes to a mind for example on a $4\times 4$ lattice Ising gauge theory has 32 degrees of freedom while Ising has 16, so somehow I would need to project out non-physical degrees from Ising gauge but I am not sure how to impose that projection with RBM. Any other ides for simulating gauge theory or how to fix this is welcome, thanks!

**Comment by gcarleo** [comment, marked as answer]
gauging out the degrees of freedom is non trivial in general. you need to adopt a suitable neural network, see for example https://journals.aps.org/prl/abstract/10.1103/PhysRevLett.127.276402

**Reply by abukva** [reply]
Thanks, I've looked at the paper and in the Figure 2. you compare equivariant, invaraint and RBM for the exactly same Hamiltonian that I want to simulate. How did you adapt RBM in order to simulate that?

**Reply by gcarleo** [reply]
please read the paper, the paper is essentially all about doing this generalization you are looking for
